File: scripts/benchmark_rdm_calculators_rsatoolbox.R
#!/usr/bin/env Rscript

# Benchmarks ordinary RDM calculators against vendored rsatoolbox.
#
# The report separates two comparisons:
# 1. Public calculator path: rMVPA condition averaging + pairwise_dist() vs
#    rsatoolbox calc_rdm(..., descriptor=) for correlation and Euclidean RDMs.
# 2. Formula path: matched R and NumPy formulas on precomputed condition means,
#    including Mahalanobis with an explicit supplied precision matrix.

thread_env <- c("OPENBLAS_NUM_THREADS", "OMP_NUM_THREADS", "VECLIB_MAXIMUM_THREADS", "MKL_NUM_THREADS")
for (var in thread_env) {
  if (!nzchar(Sys.getenv(var))) {
    do.call(Sys.setenv, as.list(stats::setNames("1", var)))
  }
}

if (!requireNamespace("jsonlite", quietly = TRUE)) {
  stop("Package `jsonlite` is required for this benchmark.", call. = FALSE)
}
if (!requireNamespace("microbenchmark", quietly = TRUE)) {
  stop("Package `microbenchmark` is required for this benchmark.", call. = FALSE)
}

if (requireNamespace("pkgload", quietly = TRUE) && file.exists("DESCRIPTION")) {
  pkgload::load_all(".", quiet = TRUE)
} else {
  library(rMVPA)
}

repo_root <- normalizePath(".", mustWork = TRUE)
vendor_src <- file.path(repo_root, "vendor", "rsatoolbox", "src")
python_helper <- file.path(repo_root, "scripts", "benchmark_rdm_calculators_rsatoolbox.py")

if (!dir.exists(vendor_src)) {
  stop("Vendored rsatoolbox source not found at `vendor/rsatoolbox/src`.", call. = FALSE)
}
if (!file.exists(python_helper)) {
  stop("Python helper not found: ", python_helper, call. = FALSE)
}

pairwise_dist <- get("pairwise_dist", envir = asNamespace("rMVPA"))
group_means <- get("group_means", envir = asNamespace("rMVPA"))
cordist <- get("cordist", envir = asNamespace("rMVPA"))
eucdist <- get("eucdist", envir = asNamespace("rMVPA"))
rdm_vector_correlation <- get(".rdm_vector_correlation", envir = asNamespace("rMVPA"))
rdm_vector_sqeuclidean <- get(".rdm_vector_sqeuclidean", envir = asNamespace("rMVPA"))
rdm_vector_sqmahalanobis <- get(".rdm_vector_sqmahalanobis", envir = asNamespace("rMVPA"))

parse_cases <- function(spec) {
  if (!nzchar(spec)) {
    return(NULL)
  }
  pieces <- strsplit(spec, ",", fixed = TRUE)[[1]]
  lapply(pieces, function(piece) {
    vals <- as.integer(strsplit(piece, "x", fixed = TRUE)[[1]])
    if (length(vals) != 3 || anyNA(vals) || any(vals <= 0)) {
      stop("Invalid case spec `", piece, "`. Expected KxVxreps, e.g. 24x128x4.",
           call. = FALSE)
    }
    list(K = vals[[1]], V = vals[[2]], reps = vals[[3]])
  })
}

default_cases <- function(profile) {
  switch(profile,
    quick = list(
      list(K = 12L, V = 64L, reps = 3L),
      list(K = 24L, V = 128L, reps = 4L),
      list(K = 40L, V = 200L, reps = 4L)
    ),
    stress = list(
      list(K = 24L, V = 128L, reps = 6L),
      list(K = 48L, V = 256L, reps = 5L),
      list(K = 80L, V = 512L, reps = 4L)
    ),
    stop("Unknown RMVPA_RDM_BENCH_PROFILE `", profile, "`. Use `quick` or `stress`.",
         call. = FALSE)
  )
}

make_case_id <- function(case) {
  paste0("K", case$K, "_V", case$V, "_R", case$reps)
}

make_bench_data <- function(K, V, reps) {
  n_obs <- K * reps
  measurements <- matrix(NA_real_, nrow = n_obs, ncol = V)
  cond <- character(n_obs)
  vox <- seq_len(V)

  row <- 1L
  for (rep in seq_len(reps)) {
    for (k in seq_len(K)) {
      measurements[row, ] <- sin(0.07 * k * vox + 0.11 * rep) +
        cos(0.03 * rep * vox + 0.17 * k) +
        0.02 * rep -
        0.015 * k +
        0.001 * (k %% 3L) * vox / V
      cond[[row]] <- sprintf("C%03d", k)
      row <- row + 1L
    }
  }

  colnames(measurements) <- sprintf("V%03d", seq_len(V))
  levels <- sprintf("C%03d", seq_len(K))
  list(
    measurements = measurements,
    cond = factor(cond, levels = levels),
    levels = levels
  )
}

make_precision <- function(V) {
  idx <- seq_len(V)
  diag_vals <- 1 + (idx %% 7L) / 10
  u <- sin(0.17 * idx)
  diag(diag_vals, nrow = V, ncol = V) + 0.02 * tcrossprod(u) / V
}

condition_means <- function(measurements, cond, levels) {
  means <- group_means(measurements, margin = 1, group = factor(cond, levels = levels))
  means <- means[levels, , drop = FALSE]
  rownames(means) <- levels
  means
}

rdm_vec <- function(dmat) {
  unname(dmat[lower.tri(dmat)])
}

rdm_correlation_formula <- function(means) {
  rdm_vector_correlation(means, method = "pearson")
}

rdm_euclidean_formula <- function(means) {
  rdm_vector_sqeuclidean(means, normalize_by_features = TRUE)
}

rdm_mahalanobis_formula <- function(means, precision) {
  rdm_vector_sqmahalanobis(means, precision, normalize_by_features = TRUE)
}

rmvpa_pairwise_correlation <- function(measurements, cond, levels) {
  means <- condition_means(measurements, cond, levels)
  rdm_vec(pairwise_dist(cordist(method = "pearson"), means))
}

rmvpa_pairwise_euclidean <- function(measurements, cond, levels) {
  means <- condition_means(measurements, cond, levels)
  rdm_vec(pairwise_dist(eucdist(), means)^2) / ncol(means)
}

summarise_microbenchmark <- function(mb, case) {
  df <- as.data.frame(mb)
  split_ms <- split(df$time / 1e6, as.character(df$expr))
  out <- do.call(rbind, lapply(names(split_ms), function(expr) {
    vals <- split_ms[[expr]]
    distance <- sub("^.*_(correlation|euclidean|mahalanobis)$", "\\1", expr)
    method <- sub("_(correlation|euclidean|mahalanobis)$", "", expr)
    data.frame(
      case_id = make_case_id(case),
      method = method,
      distance = distance,
      K = case$K,
      V = case$V,
      reps = case$reps,
      iterations = length(vals),
      median_ms = median(vals),
      min_ms = min(vals),
      max_ms = max(vals),
      mean_ms = mean(vals),
      stringsAsFactors = FALSE
    )
  }))
  rownames(out) <- NULL
  out
}

bench_rmvpa_case <- function(case, times) {
  dat <- make_bench_data(case$K, case$V, case$reps)
  means <- condition_means(dat$measurements, dat$cond, dat$levels)
  precision <- make_precision(case$V)

  r_distances <- list(
    correlation = list(
      rmvpa_pairwise_dist = rmvpa_pairwise_correlation(dat$measurements, dat$cond, dat$levels),
      rmvpa_r_formula = rdm_correlation_formula(means)
    ),
    euclidean = list(
      rmvpa_pairwise_dist = rmvpa_pairwise_euclidean(dat$measurements, dat$cond, dat$levels),
      rmvpa_r_formula = rdm_euclidean_formula(means)
    ),
    mahalanobis = list(
      rmvpa_r_formula = rdm_mahalanobis_formula(means, precision)
    )
  )

  mb <- microbenchmark::microbenchmark(
    rmvpa_pairwise_dist_correlation = rmvpa_pairwise_correlation(dat$measurements, dat$cond, dat$levels),
    rmvpa_pairwise_dist_euclidean = rmvpa_pairwise_euclidean(dat$measurements, dat$cond, dat$levels),
    rmvpa_r_formula_correlation = rdm_correlation_formula(means),
    rmvpa_r_formula_euclidean = rdm_euclidean_formula(means),
    rmvpa_r_formula_mahalanobis = rdm_mahalanobis_formula(means, precision),
    times = times
  )

  list(
    summary = summarise_microbenchmark(mb, case),
    distances = r_distances
  )
}

extract_py_distances <- function(py_cases, idx) {
  dissimilarities <- py_cases$dissimilarities
  if (is.matrix(dissimilarities)) {
    return(as.numeric(dissimilarities[idx, ]))
  }
  as.numeric(unlist(dissimilarities[[idx]], use.names = FALSE))
}

get_py_distances <- function(py_cases, case_id, method, distance) {
  idx <- which(py_cases$case_id == case_id &
                 py_cases$method == method &
                 py_cases$distance == distance)
  if (length(idx) != 1L) {
    stop("Expected one Python result for case=", case_id,
         " method=", method, " distance=", distance,
         "; got ", length(idx), ".", call. = FALSE)
  }
  extract_py_distances(py_cases, idx)
}

profile <- Sys.getenv("RMVPA_RDM_BENCH_PROFILE", "quick")
cases <- parse_cases(Sys.getenv("RMVPA_RDM_BENCH_CASES", ""))
if (is.null(cases)) {
  cases <- default_cases(profile)
}
for (i in seq_along(cases)) {
  cases[[i]]$case_id <- make_case_id(cases[[i]])
}

times <- as.integer(Sys.getenv("RMVPA_RDM_BENCH_TIMES", "10"))
if (is.na(times) || times < 1L) {
  stop("RMVPA_RDM_BENCH_TIMES must be a positive integer.", call. = FALSE)
}
warmup <- as.integer(Sys.getenv("RMVPA_RDM_BENCH_WARMUP", "2"))
if (is.na(warmup) || warmup < 0L) {
  stop("RMVPA_RDM_BENCH_WARMUP must be a non-negative integer.", call. = FALSE)
}

cat("Benchmarking ordinary RDM calculators against vendored rsatoolbox\n")
cat("Profile:", profile, " Iterations:", times, " Warmup:", warmup, "\n")
cat("Cases:", paste(vapply(cases, `[[`, character(1), "case_id"), collapse = ", "), "\n\n")

r_results <- vector("list", length(cases))
for (i in seq_along(cases)) {
  cat("Running rMVPA:", cases[[i]]$case_id, "\n")
  r_results[[i]] <- bench_rmvpa_case(cases[[i]], times)
}

config_file <- tempfile("rsatoolbox-rdm-bench-config-", fileext = ".json")
python_out <- tempfile("rsatoolbox-rdm-bench-output-", fileext = ".json")
jsonlite::write_json(
  list(vendor_src = vendor_src, times = times, warmup = warmup, cases = cases),
  config_file,
  auto_unbox = TRUE
)

python_bin <- Sys.getenv("RMVPA_PYTHON", "python")
cat("Running rsatoolbox via", python_bin, "\n")
status <- system2(
  python_bin,
  c(python_helper, "--config", config_file, "--output", python_out),
  stdout = TRUE,
  stderr = TRUE
)
exit_status <- attr(status, "status")
if (!is.null(exit_status) && exit_status != 0L) {
  cat(paste(status, collapse = "\n"), "\n")
  stop("rsatoolbox benchmark helper failed with status ", exit_status, call. = FALSE)
}
if (length(status) > 0L) {
  cat(paste(status, collapse = "\n"), "\n")
}

py_payload <- jsonlite::read_json(python_out, simplifyVector = TRUE)
py_cases <- py_payload$cases
py_summary <- py_cases[, setdiff(names(py_cases), "dissimilarities"), drop = FALSE]

parity_specs <- data.frame(
  distance = c("correlation", "correlation", "euclidean", "euclidean",
               "mahalanobis", "mahalanobis"),
  rmvpa_method = c("rmvpa_pairwise_dist", "rmvpa_r_formula",
                   "rmvpa_pairwise_dist", "rmvpa_r_formula",
                   "rmvpa_r_formula", "rmvpa_r_formula"),
  rsatoolbox_method = c("rsatoolbox_public", "rsatoolbox_numpy_formula",
                        "rsatoolbox_public", "rsatoolbox_numpy_formula",
                        "rsatoolbox_public", "rsatoolbox_numpy_formula"),
  comparison = c("public", "formula", "public", "formula",
                 "supplied_precision_public", "formula"),
  stringsAsFactors = FALSE
)

parity <- do.call(rbind, lapply(seq_along(cases), function(i) {
  case_id <- cases[[i]]$case_id
  do.call(rbind, lapply(seq_len(nrow(parity_specs)), function(j) {
    spec <- parity_specs[j, ]
    r_vec <- r_results[[i]]$distances[[spec$distance]][[spec$rmvpa_method]]
    py_vec <- get_py_distances(py_cases, case_id, spec$rsatoolbox_method, spec$distance)
    data.frame(
      case_id = case_id,
      distance = spec$distance,
      comparison = spec$comparison,
      rmvpa_method = spec$rmvpa_method,
      rsatoolbox_method = spec$rsatoolbox_method,
      max_abs_diff = max(abs(r_vec - py_vec)),
      stringsAsFactors = FALSE
    )
  }))
}))

if (any(parity$max_abs_diff > 1e-9)) {
  print(parity)
  stop("Parity check failed before speed comparison.", call. = FALSE)
}

r_summary <- do.call(rbind, lapply(r_results, `[[`, "summary"))
all_summary <- rbind(
  r_summary,
  py_summary[, names(r_summary), drop = FALSE]
)

all_summary$public_api_speedup_vs_rsatoolbox <- NA_real_
all_summary$formula_speedup_vs_numpy_reference <- NA_real_

for (case_id in unique(all_summary$case_id)) {
  for (distance in c("correlation", "euclidean")) {
    public_ref <- all_summary$median_ms[
      all_summary$case_id == case_id &
        all_summary$distance == distance &
        all_summary$method == "rsatoolbox_public"
    ]
    public_rows <- all_summary$case_id == case_id &
      all_summary$distance == distance &
      all_summary$method %in% c("rmvpa_pairwise_dist", "rsatoolbox_public")
    all_summary$public_api_speedup_vs_rsatoolbox[public_rows] <-
      public_ref / all_summary$median_ms[public_rows]
  }

  for (distance in c("correlation", "euclidean", "mahalanobis")) {
    formula_ref <- all_summary$median_ms[
      all_summary$case_id == case_id &
        all_summary$distance == distance &
        all_summary$method == "rsatoolbox_numpy_formula"
    ]
    formula_rows <- all_summary$case_id == case_id &
      all_summary$distance == distance &
      all_summary$method %in% c("rmvpa_r_formula", "rsatoolbox_numpy_formula")
    all_summary$formula_speedup_vs_numpy_reference[formula_rows] <-
      formula_ref / all_summary$median_ms[formula_rows]
  }
}

all_summary <- all_summary[order(all_summary$case_id, all_summary$distance, all_summary$method), ]
rownames(all_summary) <- NULL

public_summary <- all_summary[
  all_summary$distance %in% c("correlation", "euclidean") &
    all_summary$method %in% c("rmvpa_pairwise_dist", "rsatoolbox_public"),
]
formula_summary <- all_summary[
  all_summary$method %in% c("rmvpa_r_formula", "rsatoolbox_numpy_formula"),
]
mahalanobis_public_reference <- all_summary[
  all_summary$distance == "mahalanobis" &
    all_summary$method == "rsatoolbox_public",
]

cat("\nParity max abs diff:\n")
print(parity, row.names = FALSE)

cat("\nComparison scope:\n")
cat("- Public calculator path: `rmvpa_pairwise_dist` averages raw observations with rMVPA `group_means()` and then uses `pairwise_dist()`; `rsatoolbox_public` starts from a Dataset and uses calc_rdm(..., descriptor='cond').\n")
cat("- Euclidean parity uses rsatoolbox's convention: squared Euclidean distance divided by feature count.\n")
cat("- Mahalanobis uses an explicit supplied precision matrix and is compared as a formula/reference path because rMVPA's current `mahadist()` estimates shrinkage precision internally.\n")
cat("- The vendored rsatoolbox Cython engine is stubbed only to allow imports; these balanced RDM calculator paths do not depend on the unbuilt engine.\n")

public_cols <- c("case_id", "distance", "method", "K", "V", "reps", "iterations",
                 "median_ms", "min_ms", "max_ms", "mean_ms",
                 "public_api_speedup_vs_rsatoolbox")
formula_cols <- c("case_id", "distance", "method", "K", "V", "reps", "iterations",
                  "median_ms", "min_ms", "max_ms", "mean_ms",
                  "formula_speedup_vs_numpy_reference")
mahal_cols <- c("case_id", "distance", "method", "K", "V", "reps", "iterations",
                "median_ms", "min_ms", "max_ms", "mean_ms")

cat("\nPublic calculator timing (milliseconds; speedup > 1 means faster than rsatoolbox public API):\n")
print(public_summary[, public_cols, drop = FALSE], row.names = FALSE, digits = 4)

cat("\nFormula timing (milliseconds; speedup > 1 means faster than the NumPy formula reference):\n")
print(formula_summary[, formula_cols, drop = FALSE], row.names = FALSE, digits = 4)

cat("\nrsatoolbox public Mahalanobis timing with supplied precision (reference only; no current rMVPA public precision path):\n")
print(mahalanobis_public_reference[, mahal_cols, drop = FALSE], row.names = FALSE, digits = 4)

out_csv <- Sys.getenv("RMVPA_RDM_BENCH_OUT", "")
if (nzchar(out_csv)) {
  utils::write.csv(all_summary, out_csv, row.names = FALSE)
  cat("\nWrote benchmark summary to", out_csv, "\n")
}

File: scripts/benchmark_rdm_calculators_rsatoolbox.py
#!/usr/bin/env python3
"""Benchmark vendored rsatoolbox RDM calculators on deterministic fixtures."""

from __future__ import annotations

import argparse
import gc
import json
import os
import sys
import time
import types
import warnings
from pathlib import Path
from tempfile import mkdtemp

import numpy as np

warnings.filterwarnings(
    "ignore",
    message="invalid value encountered in cast",
    category=RuntimeWarning,
)


def configure_writable_caches() -> None:
    """Keep optional plotting imports from writing under the user's home dir."""
    cache_root = Path(mkdtemp(prefix="rsatoolbox-rdm-cache-"))
    mpl_dir = cache_root / "matplotlib"
    xdg_dir = cache_root / "xdg"
    mpl_dir.mkdir(parents=True, exist_ok=True)
    xdg_dir.mkdir(parents=True, exist_ok=True)
    os.environ.setdefault("MPLCONFIGDIR", str(mpl_dir))
    os.environ.setdefault("XDG_CACHE_HOME", str(xdg_dir))


def install_cengine_stub() -> None:
    """Allow rsatoolbox imports when the optional Cython engine is unbuilt."""
    sim_mod = types.ModuleType("rsatoolbox.cengine.similarity")
    sim_mod.similarity = lambda *args, **kwargs: None
    sim_mod.calc_one = lambda *args, **kwargs: None
    sim_mod.calc = lambda *args, **kwargs: None

    cengine_mod = types.ModuleType("rsatoolbox.cengine")
    cengine_mod.similarity = sim_mod.similarity
    cengine_mod.calc_one = sim_mod.calc_one
    cengine_mod.calc = sim_mod.calc

    sys.modules.setdefault("rsatoolbox.cengine", cengine_mod)
    sys.modules.setdefault("rsatoolbox.cengine.similarity", sim_mod)


def load_rsatoolbox(vendor_src: str | None):
    if vendor_src:
        sys.path.insert(0, vendor_src)
    configure_writable_caches()
    install_cengine_stub()
    from rsatoolbox.data import Dataset
    from rsatoolbox.rdm import calc_rdm

    return Dataset, calc_rdm


def make_measurements(k_cond: int, n_voxels: int, n_reps: int):
    vox = np.arange(1, n_voxels + 1, dtype=float)
    n_obs = k_cond * n_reps
    measurements = np.empty((n_obs, n_voxels), dtype=float)
    conds = []

    row = 0
    for rep in range(1, n_reps + 1):
        for cond in range(1, k_cond + 1):
            measurements[row, :] = (
                np.sin(0.07 * cond * vox + 0.11 * rep)
                + np.cos(0.03 * rep * vox + 0.17 * cond)
                + 0.02 * rep
                - 0.015 * cond
                + 0.001 * (cond % 3) * vox / n_voxels
            )
            conds.append(f"C{cond:03d}")
            row += 1

    return measurements, np.asarray(conds)


def make_condition_means(measurements: np.ndarray, k_cond: int, n_reps: int) -> np.ndarray:
    n_voxels = measurements.shape[1]
    return measurements.reshape(n_reps, k_cond, n_voxels).mean(axis=0)


def make_precision(n_voxels: int) -> np.ndarray:
    idx = np.arange(1, n_voxels + 1, dtype=float)
    diag_vals = 1.0 + (np.mod(idx, 7.0) / 10.0)
    u = np.sin(0.17 * idx)
    return np.diag(diag_vals) + 0.02 * np.outer(u, u) / n_voxels


def triu_vec(dmat: np.ndarray) -> np.ndarray:
    return dmat[np.triu_indices(dmat.shape[0], k=1)]


def rdm_correlation_formula(means: np.ndarray) -> np.ndarray:
    centered = means - means.mean(axis=1, keepdims=True)
    norms = np.sqrt(np.einsum("ij,ij->i", centered, centered))
    scaled = centered / norms[:, None]
    return triu_vec(1.0 - scaled @ scaled.T)


def rdm_euclidean_formula(means: np.ndarray) -> np.ndarray:
    sum_sq = np.sum(means**2, axis=1, keepdims=True)
    dmat = sum_sq + sum_sq.T - 2.0 * means @ means.T
    return triu_vec(dmat) / means.shape[1]


def rdm_mahalanobis_formula(means: np.ndarray, precision: np.ndarray) -> np.ndarray:
    kernel = means @ precision @ means.T
    diag = np.diag(kernel)
    dmat = diag[None, :] + diag[:, None] - 2.0 * kernel
    return triu_vec(dmat) / means.shape[1]


def summarize(times: list[float]) -> dict[str, float]:
    arr = np.asarray(times, dtype=float) * 1000.0
    return {
        "median_ms": float(np.median(arr)),
        "min_ms": float(np.min(arr)),
        "max_ms": float(np.max(arr)),
        "mean_ms": float(np.mean(arr)),
    }


def benchmark_case(case: dict, dataset_cls, calc_rdm, times: int, warmup: int):
    k_cond = int(case["K"])
    n_voxels = int(case["V"])
    n_reps = int(case["reps"])
    case_id = case["case_id"]

    measurements, conds = make_measurements(k_cond, n_voxels, n_reps)
    means = make_condition_means(measurements, k_cond, n_reps)
    precision = make_precision(n_voxels)
    dataset = dataset_cls(
        measurements=measurements,
        obs_descriptors={"cond": conds},
    )

    def public_correlation():
        rdm = calc_rdm(dataset, method="correlation", descriptor="cond")
        return np.asarray(rdm.dissimilarities).reshape(-1)

    def public_euclidean():
        rdm = calc_rdm(dataset, method="euclidean", descriptor="cond")
        return np.asarray(rdm.dissimilarities).reshape(-1)

    def public_mahalanobis():
        rdm = calc_rdm(
            dataset,
            method="mahalanobis",
            descriptor="cond",
            noise=precision,
        )
        return np.asarray(rdm.dissimilarities).reshape(-1)

    runners = [
        ("rsatoolbox_public", "correlation", public_correlation),
        ("rsatoolbox_public", "euclidean", public_euclidean),
        ("rsatoolbox_public", "mahalanobis", public_mahalanobis),
        ("rsatoolbox_numpy_formula", "correlation", lambda: rdm_correlation_formula(means)),
        ("rsatoolbox_numpy_formula", "euclidean", lambda: rdm_euclidean_formula(means)),
        (
            "rsatoolbox_numpy_formula",
            "mahalanobis",
            lambda: rdm_mahalanobis_formula(means, precision),
        ),
    ]

    distances = {
        (method, distance): runner()
        for method, distance, runner in runners
    }

    for _ in range(warmup):
        for _, _, runner in runners:
            runner()

    elapsed = {(method, distance): [] for method, distance, _ in runners}
    was_enabled = gc.isenabled()
    gc.disable()
    try:
        for _ in range(times):
            for method, distance, runner in runners:
                start = time.perf_counter()
                runner()
                elapsed[(method, distance)].append(time.perf_counter() - start)
    finally:
        if was_enabled:
            gc.enable()

    rows = []
    common = {
        "case_id": case_id,
        "K": k_cond,
        "V": n_voxels,
        "reps": n_reps,
        "iterations": times,
    }
    for method, distance, _ in runners:
        row = dict(common)
        row.update({
            "method": method,
            "distance": distance,
            "dissimilarities": distances[(method, distance)].tolist(),
        })
        row.update(summarize(elapsed[(method, distance)]))
        rows.append(row)

    return rows


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--config", required=True)
    parser.add_argument("--output", required=True)
    args = parser.parse_args()

    with open(args.config, "r", encoding="utf-8") as handle:
        config = json.load(handle)

    vendor_src = config.get("vendor_src")
    if vendor_src is not None:
        vendor_src = str(Path(vendor_src).resolve())

    dataset_cls, calc_rdm = load_rsatoolbox(vendor_src)
    results = []
    for case in config["cases"]:
        results.extend(benchmark_case(case, dataset_cls, calc_rdm,
                                      int(config["times"]), int(config["warmup"])))

    payload = {
        "toolbox": "rsatoolbox",
        "vendor_src": vendor_src,
        "cengine_stubbed": True,
        "cases": results,
    }
    with open(args.output, "w", encoding="utf-8") as handle:
        json.dump(payload, handle)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
